server/aesdsocket.c

```diff
@@ -53,39 +78,183 @@ const char* errs[] = {
+	union sigval v = {.sival_ptr = (void*) *(uint64_t*) pack};
+	sigqueue(getpid(), SIGTERM, v);
+	pthread_exit(NULL);
+}
+
+//-----client thread-----
+
+// close socket if cancelled
+static void vclose(void* fdpt) {
+	close(*(int*) fdpt);
+}
+
+static void* client_thread (void* param_v) {
+	// these must be declared before pushing the cleanup handler
+	// because it creates a scope...
+	int s;
+	char* cip;
+	struct client* param = (struct client*) param_v;
+	pthread_cleanup_push(vclose, &param->sock);
+
+	// log accepted connection
+	cip = alloca(INET_ADDRSTRLEN);
+	pthread_mutex_lock(&ntoa_lk);
+	const char* ipstr = inet_ntoa(param->addr);
+	strncpy(cip, ipstr, INET_ADDRSTRLEN);
+	pthread_mutex_unlock(&ntoa_lk);
+	syslog(LOG_INFO, "Accepted connection from %s\n", cip);
+	
+	// alloc packet buffer
+	size_t max_packet = INITIAL_MAX_PACKET;
+	char* packet = malloc(max_packet);
+	if (!packet) cleanup_thr(SRC_MALLOC);
+	pthread_cleanup_push(free, packet);
+
+	// get packet
+	char* delim = NULL;
+	size_t packet_sz = 0;
+	size_t max_read = max_packet;
+	while (!delim) {
+		assert(packet_sz <= max_packet);
+		ssize_t read_sz = read(param->sock, &packet[packet_sz], max_read);	
+		if (read_sz == -1)
+			cleanup_thr(SRC_READ);
+		
+		if (read_sz == max_read) { // need to expand buffer
+			max_read = max_packet;
+			max_packet *= 2;
+			packet = realloc(packet, max_packet);
+			if (!packet) cleanup_thr(SRC_REALLOC);
+		}
+
+		delim = memchr(&packet[packet_sz], '\n', read_sz);
+		packet_sz += read_sz;
+		max_read -= read_sz;
+	}
+
+	// write packet
+	pthread_mutex_lock(&of_lk);
+	if (of_sz + packet_sz > of_memsz) { // need to expand mapping
+		void* prev_mem = of_mem;
+		size_t new_memsz = ((of_sz + packet_sz)/PAGE_SIZE + 1)*PAGE_SIZE;
+		s = ftruncate(ofd, new_memsz);
+		if (s == -1) cleanup_thr(SRC_FTRUNCATE);
+		of_mem = mremap(of_mem, of_memsz, new_memsz, MREMAP_MAYMOVE);
+		if (of_mem == MAP_FAILED) cleanup_thr(SRC_MREMAP);
+		if (of_mem != prev_mem) of_pt = (char*) of_mem + of_sz;
+		of_memsz = new_memsz;
+	}
+	memcpy(of_pt, packet, packet_sz);
+	of_sz += packet_sz;
+	of_pt += packet_sz;
+	size_t write_sz = of_sz;
+	char* src_pt = (char*) of_mem;
+	pthread_mutex_unlock(&of_lk);
+
+	// send output file to client, up to its own packet
+	ssize_t wsz = 0;
+	do {
+		write_sz -= wsz;
+		src_pt += wsz;
+		wsz = write(param->sock, of_mem, write_sz);
+	} while (wsz != -1 && write_sz > wsz);
+	if (wsz == -1) cleanup_thr(SRC_WRITE);
+
+	// close & cleanup connection
+	pthread_cleanup_pop(0);
+	free(packet);
+	pthread_cleanup_pop(0);
+	s = close(param->sock);
+	if (s == -1) cleanup(SRC_CLOSE);
+	syslog(LOG_INFO, "Closed connection from %s", cip);
+	return NULL;
+}
+
+// periodic time logger (SIGALRM handler)
+static void wrtime (int sig) {
+	assert(sig == SIGALRM);
+	time_t t = time(NULL);
+	size_t s = strftime((char*) &timestr, MAX_TIMELEN, "timestamp:%a, %d %b %Y %T %z\n", localtime(&t));
+	if (s == 0) cleanup(SRC_STRFTIME);
+	pthread_mutex_lock(&of_lk);
```

Comment:
@georgehodgkins is this safe to do in a signal handler?  What happens if the signal for alarm occurs when you have the lock held in your thread?